lib/srv/alpnproxy/local_proxy.go

```diff
@@ -0,0 +1,288 @@
+		cancel:  cancel,
+	}, nil
+}
+
+// SSHProxy is equivalent of `ssh -o 'ForwardAgent yes' -p port  %r@host -s proxy:%h:%p` but established SSH
+// connection to RemoteProxyAddr is wrapped with TLS protocol.
+func (l *LocalProxy) SSHProxy() error {
+	upstreamConn, err := tls.Dial("tcp", l.cfg.RemoteProxyAddr, &tls.Config{
+		NextProtos:         []string{string(l.cfg.Protocol)},
+		InsecureSkipVerify: l.cfg.InsecureSkipVerify,
+		ServerName:         l.cfg.SNI,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer upstreamConn.Close()
+
+	sshAgent, err := getAgent()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	client, err := makeSSHClient(upstreamConn, l.cfg.RemoteProxyAddr, &ssh.ClientConfig{
+		User: l.cfg.SSHUser,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeysCallback(sshAgent.Signers),
+		},
+		HostKeyCallback: l.cfg.SSHHostKeyCallback,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer client.Close()
+
+	sess, err := client.NewSession()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer sess.Close()
+
+	err = agent.ForwardToAgent(client, sshAgent)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = agent.RequestAgentForwarding(sess)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err = sess.RequestSubsystem(proxySubsystemName(l.cfg.SSHUserHost)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := proxySession(l.context, sess); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func proxySubsystemName(userHost string) string {
+	return fmt.Sprintf("proxy:%s", userHost)
+}
+
+func makeSSHClient(conn *tls.Conn, addr string, cfg *ssh.ClientConfig) (*ssh.Client, error) {
+	cc, chs, reqs, err := ssh.NewClientConn(conn, addr, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return ssh.NewClient(cc, chs, reqs), nil
+}
+
+func proxySession(ctx context.Context, sess *ssh.Session) error {
+	stdout, err := sess.StdoutPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stdin, err := sess.StdinPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stderr, err := sess.StderrPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	errC := make(chan error)
+	go func() {
+		_, err := io.Copy(os.Stdout, stdout)
+		errC <- err
+	}()
+	go func() {
+		_, err := io.Copy(stdin, os.Stdin)
+		errC <- err
+	}()
+	go func() {
+		_, err := io.Copy(os.Stderr, stderr)
+		errC <- err
+	}()
+	var errs []error
+	for i := 0; i < 3; i++ {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errC:
+			if err != nil && !errors.Is(err, io.EOF) {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// Start starts the LocalProxy.
+func (l *LocalProxy) Start(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		conn, err := l.cfg.Listener.Accept()
+		if err != nil {
+			if utils.IsOKNetworkError(err) {
+				return nil
+			}
+			log.WithError(err).Errorf("Faield to accept client connection.")
+			continue
+		}
```

Comment:
If you just continue on an unknown error from `Listener.Accept()`, is there a chance the Listener is in a bad state and every future `Accept` call will return an error, and this will just be a hot loop until the context expires?